samples/net/mqtt/src/modules/transport/transport.c

```diff
@@ -0,0 +1,308 @@
+{
+	int err;
+	struct mqtt_publish_param param = {
+		.message.payload.data = payload->string,
+		.message.payload.len = strlen(payload->string),
+		.message.topic.qos = MQTT_QOS_1_AT_LEAST_ONCE,
+		.message_id = k_uptime_get_32(),
+		.message.topic.topic.utf8 = CONFIG_MQTT_SAMPLE_TRANSPORT_PUBLISH_TOPIC,
+		.message.topic.topic.size = strlen(CONFIG_MQTT_SAMPLE_TRANSPORT_PUBLISH_TOPIC),
+	};
+
+	err = mqtt_helper_publish(&param);
+	if (err) {
+		LOG_ERR("Failed to send payload, err: %d", err);
+		return;
+	}
+
+	LOG_INF("Publishing message: \"%.*s\" on topic: \"%.*s\"", param.message.payload.len,
+								   param.message.payload.data,
+								   param.message.topic.topic.size,
+								   param.message.topic.topic.utf8);
+}
+
+static void subscribe(void)
+{
+	int err;
+	struct mqtt_topic topics[] = {
+		{
+			.topic.utf8 = CONFIG_MQTT_SAMPLE_TRANSPORT_SUBSCRIBE_TOPIC,
+			.topic.size = strlen(CONFIG_MQTT_SAMPLE_TRANSPORT_SUBSCRIBE_TOPIC),
+		},
+	};
+	struct mqtt_subscription_list list = {
+		.list = topics,
+		.list_count = ARRAY_SIZE(topics),
+		.message_id = SUBSCRIBE_TOPIC_ID,
+	};
+
+	err = mqtt_helper_subscribe(&list);
+	if (err) {
+		LOG_ERR("Failed to subscribe to topics, error: %d", err);
+		return;
+	}
+}
+
+/* Connect work - Used to establish a connection to the MQTT broker and schedule reconnection
+ * attempts.
+ */
+static void connect_work_fn(struct k_work *work)
+{
+	ARG_UNUSED(work);
+
+	__ASSERT(client_id_get(client_id, sizeof(client_id)) == 0, "Failed getting Client ID");
+
+	struct mqtt_helper_conn_params conn_params = {
+		.hostname.ptr = CONFIG_MQTT_SAMPLE_TRANSPORT_BROKER_HOSTNAME,
+		.hostname.size = strlen(CONFIG_MQTT_SAMPLE_TRANSPORT_BROKER_HOSTNAME),
+		.device_id.ptr = client_id,
+		.device_id.size = strlen(client_id),
+	};
+
+	int err = mqtt_helper_connect(&conn_params);
+
+	if (err) {
+		LOG_ERR("Failed connecting to MQTT, error code: %d", err);
+	}
+
+	k_work_reschedule(&connect_work,
+			  K_SECONDS(CONFIG_MQTT_SAMPLE_TRANSPORT_RECONNECTION_TIMEOUT_SECONDS));
+}
+
+/* Zephyr State Machine framework handlers */
+static void disconnected_entry(void *o)
+{
+	struct s_object *od = o;
+
+	/* Reschedule a connection attempt if we are connected to network and we enter the
+	 * disconnected state.
+	 */
+	if (od->status == NETWORK_CONNECTED) {
+		k_work_reschedule_for_queue(&queue, &connect_work, K_NO_WAIT);
+	}
+}
+
+static void disconnected_run(void *o)
+{
+	struct s_object *od = o;
+
+	if ((od->status == NETWORK_CONNECTED) && (od->chan == &NETWORK_CHAN)) {
+
+		/* Wait for 5 seconds to ensure that the network stack is ready before
+		 * attempting to connect to MQTT. This delay is only needed when building for
+		 * Wi-Fi.
+		 */
+		k_work_reschedule_for_queue(&queue, &connect_work, K_SECONDS(5));
+	}
+}
+
+static void connected_entry(void *o)
+{
+	LOG_INF("Connected to MQTT broker");
+	LOG_INF("Hostname: %s", CONFIG_MQTT_SAMPLE_TRANSPORT_BROKER_HOSTNAME);
+	LOG_INF("Client ID: %s", client_id);
```

Comment:
Could we print the Client ID before connecting? We were just debugging a broken config and that would have been helpful.

Reply:
Good idea! Ill try to add it before release. But, first and foremost ill try to get this sample in asap.